dlt/common/libs/pyiceberg.py

```diff
@@ -0,0 +1,150 @@
+from typing import Dict, Any, List
+import os
+
+from dlt import version, Pipeline
+from dlt.common.libs.pyarrow import cast_arrow_schema_types, columns_to_arrow
+from dlt.common.schema.typing import TWriteDisposition
+from dlt.common.utils import assert_min_pkg_version
+from dlt.common.exceptions import MissingDependencyException
+from dlt.common.configuration.specs import CredentialsConfiguration
+from dlt.common.configuration.specs.mixins import WithPyicebergConfig
+from dlt.destinations.impl.filesystem.filesystem import FilesystemClient
+
+assert_min_pkg_version(
+    pkg_name="sqlalchemy",
+    version="2.0.18",
+    msg="`sqlalchemy>=2.0.18` is needed for `iceberg` table format on `filesystem` destination.",
+)
+
+try:
+    from pyiceberg.table import Table as IcebergTable
+    from pyiceberg.catalog.sql import SqlCatalog
+    import pyarrow as pa
+except ModuleNotFoundError:
+    raise MissingDependencyException(
+        "dlt pyiceberg helpers",
+        [f"{version.DLT_PKG_NAME}[pyiceberg]"],
+        "Install `pyiceberg` so dlt can create Iceberg tables in the `filesystem` destination.",
+    )
+
+
+DLT_ICEBERG_NAMESPACE = "dlt"
+
+
+def ensure_iceberg_compatible_arrow_schema(schema: pa.Schema) -> pa.Schema:
+    ARROW_TO_ICEBERG_COMPATIBLE_ARROW_TYPE_MAP = {
+        pa.types.is_time: pa.string(),
+        pa.types.is_decimal256: pa.string(),  # pyarrow does not allow downcasting to decimal128
+    }
+    return cast_arrow_schema_types(schema, ARROW_TO_ICEBERG_COMPATIBLE_ARROW_TYPE_MAP)
+
+
+def ensure_iceberg_compatible_arrow_data(data: pa.Table) -> pa.Table:
+    schema = ensure_iceberg_compatible_arrow_schema(data.schema)
+    return data.cast(schema)
+
+
+def write_iceberg_table(
+    table: IcebergTable,
+    data: pa.Table,
+    write_disposition: TWriteDisposition,
+) -> None:
+    if write_disposition == "append":
+        table.append(ensure_iceberg_compatible_arrow_data(data))
+    elif write_disposition == "replace":
+        table.overwrite(ensure_iceberg_compatible_arrow_data(data))
+
+
+def get_catalog(
+    client: FilesystemClient,
+    table_name: str,
+    schema: pa.Schema = None,
+    partition_columns: List[str] = None,
+) -> SqlCatalog:
+    """Returns single-table, ephemeral, in-memory Iceberg catalog."""
+
+    # create in-memory catalog
+    catalog = SqlCatalog(
+        "default",
+        uri="sqlite:///:memory:",
+        **_get_fileio_config(client.config.credentials),
+    )
+    catalog.create_namespace(DLT_ICEBERG_NAMESPACE)
+
+    # add table to catalog
+    table_id = f"{DLT_ICEBERG_NAMESPACE}.{table_name}"
+    table_path = f"{client.dataset_path}/{table_name}"
+    metadata_path = f"{table_path}/metadata"
+    if client.fs_client.exists(metadata_path):
```

Comment:
is it possible to split the function that returns a catalog (even per table so a new catalog is simply empty) and function that does table maintenance (evolves schema or creates empty table)

IMO the code structure would be
```py
with get_catalog() as catalog:
   create_or_evolve_table()
   write_data()
```

Reply:
where catalog is abstract catalog class of pyiceberg